### PR TITLE
DateTimeInput POC

### PR DESCRIPTION
**Pull request by sanmiawi**
Integrates the initial DateTime component with react datepicker component

### CODE REVIEW

File: pyrene/src/components/DateTimePicker/RangeDateTimeInput/RangeDateTimeInput.tsx [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+  endTimeValue,
+  startDateValue,
+  startTimeValue,
+  timeStamps,
+  timeZone = 'Europe/Zurich',
+  maxDateTime = getFutureDate({ years: 1 }),
+  minDateTime = 0,
+}: RangeProps) => {
+
+  const [iendDate, setEndDate] = useState<Date | undefined>(undefined);
```
Comment (Sakai-san) [inline pyrene/src/components/DateTimePicker/RangeDateTimeInput/RangeDateTimeInput.tsx:86]: When you have plenty of state like you have here, it s recommended to use useReducer. I used it here.

**Comment by justmejulian** [comment]
Please also add me as a Reviewer once this is done/ready to review.

File: pyrene/src/utils/DateUtils.ts [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+
+export const standardEUDateFormat = (date: Date): string => format(date, 'dd.MM.yyyy');
+
+export const standardEUTimeFormat = (date: Date): string => format(date, 'HH:mm');
+
+export const getDateTypeFromddmmyyyyWithSep = (str: string): DateType | undefined => {
+  if (str.length === 10 && allowedSeparatorCheck(str.charAt(2)) && allowedSeparatorCheck(str.charAt(5))) {
+    const date = { day: +str.substr(0, 2), month: +str.substr(3, 2), year: +str.substr(6) };
+    if (!Number.isNaN(date.day) && !Number.isNaN(date.month) && !Number.isNaN(date.year)) {
+      return date;
```
Comment (Sakai-san) [inline pyrene/src/utils/DateUtils.ts:189]: Two remarks about the function `getDateTypeFromddmmyyyyWithSep`
1. This function `parses` a date in form of a string
2. This function depends on another function called `allowedSeparatorCheck`.

**My advices**
1. Use an established `parser` for that, like this

```
import { parse, getDay, getMonth, getYear, isValid } from 'date-fns';

const getDateTypeFromddmmyyyyWithSep = (dateString: string): DateType | undefined => {
    const parsed = parse(dateString, 'dd.MM.yyyy hh:mm', new Date());
    return isValid(parsed) ? { day: getDay(parsed), month: getMonth(parsed), year: getYear(parsed) } : undefined;
};

```

2. Use the `Dependency Injection` pattern to pass a required function in the body of the main function.

File: pyrene/src/components/DateTimePicker/RangeSelector/DetectClickOutside.tsx [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+      document.addEventListener('keydown', onKeyDn);
+
+      return () => {
+        // Remove listeners
+        document.removeEventListener('mousedown', handleEvent, false);
+        document.removeEventListener('touchend', handleEvent, false);
+        document.removeEventListener('keydown', onKeyDn);
+      };
+    }
+    return () => {};
```
Comment (Sakai-san) [inline pyrene/src/components/DateTimePicker/RangeSelector/DetectClickOutside.tsx:56]: This line of code is not necessary.

File: pyrene/src/components/DateTimePicker/DateTimePicker.tsx [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+  onCalendarOpen,
+  onChange,
+  onClickOutside,
+  range = false,
+  selectEnd,
+  selectStart,
+  startDate,
+  tabNum,
+  timeStamp,
+  timeZone = 'Europe/Zurich',
```
Comment (Sakai-san) [inline pyrene/src/components/DateTimePicker/DateTimePicker.tsx:109]: Use the [default](https://stackoverflow.com/questions/6939685/get-client-time-zone-from-browser) time zone of user's web client.

File: pyrene/src/components/DateTimePicker/RangeSelector/RangeSelector.tsx [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+}
+
+const RangeSelector: React.FC<RangeSelectorProps> = (({
+  dateOnly = false,
+  maxDateTime = getFutureDate({ years: 1 }),
+  minDateTime = 0,
+  labels = ['From', 'To'],
+  onChange,
+  timeStamps,
+  timeZone = 'Europe/Zurich',
```
Comment (Sakai-san) [inline pyrene/src/components/DateTimePicker/RangeSelector/RangeSelector.tsx:66]: Use the [default](https://stackoverflow.com/questions/6939685/get-client-time-zone-from-browser) local time zone of user's web client.

File: pyrene/src/utils/DateUtils.ts [hunk 6]
```diff
@@ -0,0 +1,264 @@
+import {
```
Comment (Sakai-san) [inline pyrene/src/utils/DateUtils.ts:1]: [Performance](https://stackoverflow.com/questions/38171465/modules-import-and-destructuring-performances) on import.
Doing so is more performant :

```
import  isExists from 'date-fns/isExists';
import  sub from 'date-fns/sub';
import  add from 'date-fns/add';
import  Duration from 'date-fns/Duration';
import  format from 'date-fns/format';
```

File: pyrene/src/utils/DateUtils.ts [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+
+import {
+  zonedTimeToUtc, utcToZonedTime,
+} from 'date-fns-tz';
+
+const DAY = 'day';
+const MONTH = 'month';
+const YEAR = 'year';
+
+export const DATE_UNITS = {
```
Comment (Sakai-san) [inline pyrene/src/utils/DateUtils.ts:13]: This is a use for a [ENUM](https://www.typescriptlang.org/docs/handbook/enums.html) :

```
export enum DATE_UNITS  {
  DAY,
  MONTH,
  YEAR,
};

```

File: pyrene/src/components/DateTimePicker/RangeSelector/DetectClickOutside.tsx (hunk 3 above)
Comment (sanmiawi) [reply pyrene/src/components/DateTimePicker/RangeSelector/DetectClickOutside.tsx:56]: If I remove this line, then I get linting error. I think this way is cleaner than disabling the linting rule but if for any reason it is better to remove it and disable the rule please let me know.

File: pyrene/src/utils/DateUtils.ts [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+  return `${hours}:${minutes}`;
+};
+
+export const dateTypeToStandardEUDateFormat = (date: DateType): string => {
+  const day = zeroFill(date.day.toString(), 2);
+  const month = zeroFill(date.month.toString(), 2);
+  const year = zeroFill(date.year.toString(), 4);
+
+  return `${day}.${month}.${year}`;
+};
```
Comment (Sakai-san) [inline pyrene/src/utils/DateUtils.ts:179]: Remarks about the function `dateTypeToStandardEUDateFormat`

1. This function format a string
2. This function depends on another function called `zeroFill`.
3. Naming

**My advices**

1. Use an established parser for that, like this
```
 const customDateFormat = (date: string, pattern: string): string => {
    const parsed = parse(date, pattern, new Date());
    return format(parsed,  pattern);
  };
  
// usage 
const dateType = { day: 22, month: 7, year: 2021 };
const formatedEU = customDateFormat( `${dateType.day}.${dateType.month}.${dateType.year}`, 'dd.MM.yyyy' ));
const formatedUS = customDateFormat( `${dateType.year}.${dateType.month}.${dateType.day}, 'yyyy-MM-dd' ));
```


2. Use the `Dependency Injection` pattern to pass a required function in the body of the main function.

3. `Naming` is important for your code reviewer and reader. `customDateFormat` might be easier to read.

4. Do apply the `DRY` principle

File: pyrene/src/components/DateTimePicker/DateTimeInput/DateTimeInput.tsx [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+        setDateValue(node.value);
+        setTimeValue('');
+
+        return onChange(event);
+      }
+    }
+    return null;
+  };
+
+  const formatTime = (value: string) => {
```
Comment (Sakai-san) [inline pyrene/src/components/DateTimePicker/DateTimeInput/DateTimeInput.tsx:69]: I do not get what the `formatTime` does.

File: pyrene/src/utils/DateUtils.ts [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+export const errorDateBool = (datestring: string, minimumValue?: number, maximumValue?: number, timeZone?: string): boolean => {
+  if (datestring.length <= 10) {
+    if ((datestring.length < 10 || isValidDate(getDateTypeFromddmmyyyyWithSep(datestring))) && getErrors(false, false, datestring, minimumValue, maximumValue, timeZone) === '') {
+      return false;
+    }
+  }
+  return true;
+};
+
+export const errorTimeBool = (timestring: string): boolean => {
```
Comment (Sakai-san) [inline pyrene/src/utils/DateUtils.ts:257]: Naming is better so `hasTimeError`

File: pyrene/src/utils/DateUtils.ts [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+    return 'Invalid date format';
+  }
+  if (timeInvalid) {
+    return 'Invalid time format';
+  }
+  return '';
+};
+
+
+export const errorDateBool = (datestring: string, minimumValue?: number, maximumValue?: number, timeZone?: string): boolean => {
```
Comment (Sakai-san) [inline pyrene/src/utils/DateUtils.ts:248]: Namming is better so `hasErrorDate`

File: pyrene/src/utils/DateUtils.ts [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+  if (timestampToCheck < minimumValue) {
+    return -1;
+  }
+  if (timestampToCheck > maximumValue) {
+    return 1;
+  }
+  return 0;
+};
+
+export const getErrors = (dateInvalid?: boolean, timeInvalid?: boolean, dateString?: string, minimumValue?: number, maximumValue?: number, timeZone?: string): string => {
```
Comment (Sakai-san) [inline pyrene/src/utils/DateUtils.ts:215]: `getErrors` is very difficult to read. If the function take more than 3 params, it s [better](https://eslint.org/docs/rules/max-params) to pass an object.

File: pyrene/src/components/DateTimePicker/DateTimePicker.tsx [hunk 13]
```diff
@@ -0,0 +1,290 @@
...
+      setInternalDate(undefined);
+
+      if (onFunction) {
+        onFunction(undefined);
+      }
+    }
+  }, [dateOnly, timeZone]);
+
+  // Handle changes from react datepicker
+  const onChangeReactDP = (date: Date | [Date, Date] | null, event: React.SyntheticEvent<any> | undefined): void => {
```
Comment (Sakai-san) [inline pyrene/src/components/DateTimePicker/DateTimePicker.tsx:149]: `onChangeReactDP` is 50 lines of code long. Quite hard to follow,

File: pyrene/src/components/DateTimePicker/DateTimeInput/DateTimeInput.tsx [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+  setTimeValue = () => {},
+  tabNum,
+  timeValue = '',
+}:DateTimeInputProps, ref:React.Ref<HTMLInputElement>) => {
+
+  const handleDateOnChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    const node = event.target as HTMLInputElement;
+
+    if (allowedValueCheck(node.value)) {
+      if (!dateOnly && node.value.length > 10) {
```
Comment (Sakai-san) [inline pyrene/src/components/DateTimePicker/DateTimeInput/DateTimeInput.tsx:52]: You might have a `constant` for storing the length, like this 

```
const enum DATE_LENGTH {
  DATE_ONLY= 5;
  DATE_WITH_TIME =10,
}
```

File: pyrene/src/components/DateTimePicker/DateTimeInput/DateTimeInput.tsx (hunk 9 above)
Comment (sanmiawi) [reply pyrene/src/components/DateTimePicker/DateTimeInput/DateTimeInput.tsx:69]: This function is more for aesthetics. How it works is this: if the user forgets to put a space between the date and time it automatically inserts it. Example, if user types '11.01.200111:30' it will be separated to '11.01.2001 11:30' this is not important and can be removed.

**Review by Sakai-san — APPROVED** [review]

**Review by justmejulian — CHANGES_REQUESTED** [review]
These are just some small comments. I would like to run it and dive deep as well.

File: pyrene/src/utils/DateUtils.ts [hunk 15]
```diff
@@ -0,0 +1,264 @@
...
+  if (timeZone && !isValidTimeZone(timeZone)) {
+    return 'Invalid time zone.';
+  }
+
+  if (dateString && timeZone) {
+    const tmpDate = getDateTypeFromddmmyyyyWithSep(dateString);
+    if (tmpDate && isValidDate(tmpDate)) {
+      if (typeof minimumValue !== 'undefined' && typeof maximumValue !== 'undefined') {
```
[the remark this answers is not in the document]
Comment (sanmiawi) [reply pyrene/src/utils/DateUtils.ts:223]: If we do not use typeof then the date value does not get checked against a 0 value. The `minimumValue` can be 0 which is 01.01.1970

File: pyrene/src/utils/DateUtils.ts (hunk 8 above)
Comment (sanmiawi) [reply pyrene/src/utils/DateUtils.ts:179]: Ended up removing the zerofill function and using a `customDateFormat` function

File: pyrene/src/components/DateTimePicker/DateTimeInput/DateTimeInput.tsx (hunk 9 above)
Comment (sanmiawi) [reply pyrene/src/components/DateTimePicker/DateTimeInput/DateTimeInput.tsx:69]: ✅

File: pyrene/src/utils/DateUtils.ts (hunk 11 above)
Comment (sanmiawi) [reply pyrene/src/utils/DateUtils.ts:248]: ✅

File: pyrene/src/utils/DateUtils.ts (hunk 12 above)
Comment (sanmiawi) [reply pyrene/src/utils/DateUtils.ts:215]: ✅

File: pyrene/src/components/DateTimePicker/DateTimePicker.tsx (hunk 13 above)
Comment (sanmiawi) [reply pyrene/src/components/DateTimePicker/DateTimePicker.tsx:149]: ✅

File: pyrene/src/components/DateTimePicker/DateTimeInput/DateTimeInput.tsx (hunk 14 above)
Comment (sanmiawi) [reply pyrene/src/components/DateTimePicker/DateTimeInput/DateTimeInput.tsx:52]: ✅

File: pyrene/src/components/DateTimePicker/DateTimeInput/DateTimeInput.tsx [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+  setDateValue?: (value: string) => void,
+  setTimeValue?: (value: string) => void,
+  tabNum?: number,
+  timeValue?: string,
+  value?: string
+}
+
+const DateTimeInput = forwardRef(({
+  dateOnly = false,
+  dateValue = '',
```
[the remark this answers is not in the document]
Comment (sanmiawi) [reply pyrene/src/components/DateTimePicker/DateTimeInput/DateTimeInput.tsx:33]: There will be the need to check for `undefined` before usage in a number of places if this is implemented. I feel that could make the code harder to read

File: pyrene/src/components/DateTimePicker/DateTimePicker.tsx [hunk 17]
```diff
@@ -0,0 +1,290 @@
...
+  minDateTime?: number,
+  /**
+   * Name that can be used to uniquely identify the component
+   */
+  name?: string,
+  /**
+   * Function to handle onBlur event
+   */
+  onBlur?: () => OnFunction,
+  onCalendarOpen?: () => void,
```
[the remark this answers is not in the document]
Comment (sanmiawi) [reply pyrene/src/components/DateTimePicker/DateTimePicker.tsx:56]: ✅

File: pyrene/src/utils/DateUtils.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (sanmiawi) [reply pyrene/src/utils/DateUtils.ts:223]: ✅

File: pyrene/src/components/DateTimePicker/DateTimeInput/DateTimeInput.tsx [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+    }
+    return '';
+  };
+
+  return (
+    <div
+      className={styles.dateTimeComponent}
+      onKeyUp={() => handleOn?.(dateValue, timeValue.trim())}
+    >
+      <div className={styles.dateTimeFieldTitle}>{label || (dateOnly ? 'Date' : 'Date & Time')}</div>
```
[the remark this answers is not in the document]
Comment (sanmiawi) [reply pyrene/src/components/DateTimePicker/DateTimeInput/DateTimeInput.tsx:84]: ✅

File: pyrene/src/components/DateTimePicker/DateTimePicker.spec.tsx [hunk 19]
```diff
@@ -0,0 +1,228 @@
...
+
+    const rendered = mount(<DateTimePicker {...props} />);
+    const dateInput = rendered.find('input').first();
+    const timeInput = rendered.find('input').last();
+
+    expect(dateInput.props().placeholder?.includes('DD.MM.YYY') && dateInput.props().value === '').not.toBeFalsy();
+    expect(timeInput.props().placeholder?.includes('HH:MM') && timeInput.props().value === '').not.toBeFalsy();
+  });
+/*
+  it('Valid timestamp input calls onChange with timestamp.', () => {
```
[the remark this answers is not in the document]
Comment (sanmiawi) [reply pyrene/src/components/DateTimePicker/DateTimePicker.spec.tsx:41]: Tests are planned for the nearest future

File: pyrene/src/components/DateTimePicker/DateTimeInput/DateTimeInput.tsx [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+
+const DateTimeInput = forwardRef(({
+  dateOnly = false,
+  dateValue = '',
+  errorValue = '',
+  handleOn,
+  disabled,
+  label,
+  name = '',
+  onChange = () => {},
```
[the remark this answers is not in the document]
Comment (sanmiawi) [reply pyrene/src/components/DateTimePicker/DateTimeInput/DateTimeInput.tsx:39]: ✅

File: pyrene/src/utils/DateUtils.ts (hunk 7 above)
Comment (sanmiawi) [reply pyrene/src/utils/DateUtils.ts:13]: ✅

File: pyrene/src/utils/DateUtils.ts (hunk 6 above)
Comment (sanmiawi) [reply pyrene/src/utils/DateUtils.ts:1]: ✅

File: pyrene/src/components/DateTimePicker/RangeSelector/RangeSelector.tsx (hunk 5 above)
Comment (sanmiawi) [reply pyrene/src/components/DateTimePicker/RangeSelector/RangeSelector.tsx:66]: The decision now is that we use a time zone prop, if there is one else we get it from the browser

Comment (sanmiawi) [reply pyrene/src/components/DateTimePicker/RangeSelector/RangeSelector.tsx:66]: ✅

File: pyrene/src/components/DateTimePicker/DateTimePicker.tsx [hunk 21]
```diff
@@ -0,0 +1,290 @@
...
+  }) => React.ReactNode,
+  /**
+   * Boolean to toggle time display
+   */
+  dateOnly?: boolean,
+  /**
+   * This is a Date object that represents the end date of the component
+   */
+  endDate?: Date,
+  inline?: boolean,
```
[the remark this answers is not in the document]
Comment (sanmiawi) [reply pyrene/src/components/DateTimePicker/DateTimePicker.tsx:35]: ✅

File: pyrene/src/components/DateTimePicker/DateTimePicker.tsx [hunk 22]
```diff
@@ -0,0 +1,290 @@
...
+  selectEnd?: boolean,
+  /**
+  * Boolean to indicate if this input selects the start date of a range
+  */
+  selectStart?: boolean,
+  /**
+  * This is a Date object that represents the start date of the component
+  */
+  startDate?: Date,
+  tabNum?: number,
```
[the remark this answers is not in the document]
Comment (sanmiawi) [reply pyrene/src/components/DateTimePicker/DateTimePicker.tsx:78]: ✅

File: pyrene/src/components/DateTimePicker/DateTimePicker.tsx [hunk 23]
```diff
@@ -0,0 +1,290 @@
...
+
+  // Handle changes from react datepicker
+  const onChangeReactDP = (date: Date | [Date, Date] | null, event: React.SyntheticEvent<any> | undefined): void => {
+    if (date && (event?.type === 'click' || (event?.type === 'keydown' && (event as React.KeyboardEvent).key.length > 1))) {
+      if (!Array.isArray(date)) {
+        setDateValue(standardEUDateFormat(date));
+
+        if (dateOnly) {
+          handleOn(standardEUDateFormat(date), '00:00', onChange);
+        } else if (!dateOnly && timeValue && timeValue !== '') {
```
[the remark this answers is not in the document]
Comment (sanmiawi) [reply pyrene/src/components/DateTimePicker/DateTimePicker.tsx:156]: ✅

File: pyrene/src/utils/DateUtils.ts (hunk 2 above)
Comment (sanmiawi) [reply pyrene/src/utils/DateUtils.ts:189]: This function has been removed

File: pyrene/src/components/DateTimePicker/DateTimePicker.tsx [hunk 24]
```diff
@@ -0,0 +1,290 @@
...
+    if (typeof timeStamp === 'number') {
+      const dateObj = new Date(timeStamp);
+      if (!Number.isNaN(dateObj.valueOf())) {
+        setInternalDate(convertToZoneTime(timeStamp, timeZone));
+        setInvalidTimestamp(false);
+      } else {
+        setInternalDate(undefined);
+        setInvalidTimestamp(true);
+      }
+    } else if (typeof timeStamp === 'undefined') {
```
[the remark this answers is not in the document]
Comment (sanmiawi) [reply pyrene/src/components/DateTimePicker/DateTimePicker.tsx:224]: ✅

File: pyrene/src/components/DateTimePicker/DateTimePicker.tsx [hunk 25]
```diff
@@ -0,0 +1,290 @@
...
+   */
+  minDateTime?: number,
+  /**
+   * Name that can be used to uniquely identify the component
+   */
+  name?: string,
+  /**
+   * Function to handle onBlur event
+   */
+  onBlur?: () => OnFunction,
```
[the remark this answers is not in the document]
Comment (sanmiawi) [reply pyrene/src/components/DateTimePicker/DateTimePicker.tsx:55]: ✅

File: pyrene/src/components/DateTimePicker/DateTimePicker.tsx [hunk 26]
```diff
@@ -0,0 +1,290 @@
...
+  const [internalDate, setInternalDate] = useState<Date | undefined>();
+
+  const [dateValue, setDateValue] = useState('');
+  const [timeValue, setTimeValue] = useState('');
+
+  const [invalidTimestamp, setInvalidTimestamp] = useState(false);
+  const [errorValue, setErrorValue] = useState('');
+
+  // Sets internal date and passes validated value to parent
+  const handleOn = useCallback((dateString: string, timeString: string, onFunction?: OnFunction) => {
```
[the remark this answers is not in the document]
Comment (sanmiawi) [reply pyrene/src/components/DateTimePicker/DateTimePicker.tsx:121]: renamed✅

File: pyrene/src/components/DateTimePicker/DateTimePicker.tsx [hunk 27]
```diff
@@ -0,0 +1,290 @@
...
+  range = false,
+  selectEnd,
+  selectStart,
+  startDate,
+  tabNum,
+  timeStamp,
+  timeZone = 'Europe/Zurich',
+}: DateTimePickerProps) => {
+
+  const [internalDate, setInternalDate] = useState<Date | undefined>();
```
[the remark this answers is not in the document]
Comment (sanmiawi) [reply pyrene/src/components/DateTimePicker/DateTimePicker.tsx:112]: Yes it maintains the state of the time between the dropdown calendar and the input box. When this is not available, if a datetime has been previously selected and the dropdown is again triggered, the dropdown is not initialized with the previously selected datetime.

File: pyrene/src/components/DateTimePicker/RangeSelector/DetectClickOutside.tsx (hunk 3 above)
Comment (sanmiawi) [reply pyrene/src/components/DateTimePicker/RangeSelector/DetectClickOutside.tsx:56]: The RangeSelector component is on hold.
